File: backend/src/services/authService.ts
import { PrismaClient } from '@prisma/client';
import { hashPassword, comparePassword } from '../utils/password';
import { generateAccessToken, generateRefreshToken, verifyRefreshToken, JWTPayload } from '../utils/jwt';

const prisma = new PrismaClient();

export interface RegisterData {
  name: string;
  username: string;
  email: string;
  password: string;
}

export class AuthService {
  async register(userData: RegisterData) {
    const { name, username, email, password } = userData;

    // Check if user already exists
    const existingUser = await prisma.user.findFirst({
      where: {
        OR: [
          { email },
          { username }
        ]
      }
    });

    if (existingUser) {
      if (existingUser.email === email) {
        throw new Error('Email already exists');
      }
      if (existingUser.username === username) {
        throw new Error('Username already exists');
      }
    }

    // Hash password
    const hashedPassword = await hashPassword(password);

    // Create user
    const user = await prisma.user.create({
      data: {
        name,
        username,
        email,
        password: hashedPassword,
      },
      select: {
        id: true,
        name: true,
        username: true,
        email: true,
        created_at: true,
      }
    });

    // Generate tokens
    const tokenPayload: JWTPayload = { userId: user.id, email: user.email };
    const accessToken = generateAccessToken(tokenPayload);
    const refreshToken = generateRefreshToken(tokenPayload);

    return {
      user,
      accessToken,
      refreshToken,
    };
  }

  async login(loginIdentifier: string, password: string) {
    // Check if loginIdentifier is email or username
    const isEmail = loginIdentifier.includes('@');
    
    const user = await prisma.user.findFirst({
      where: isEmail ? { email: loginIdentifier } : { username: loginIdentifier }
    });

    if (!user) {
      throw new Error(isEmail ? 'Invalid email or password' : 'Invalid username or password');
    }

    // Verify password
    const isPasswordValid = await comparePassword(password, user.password);
    if (!isPasswordValid) {
      throw new Error(isEmail ? 'Invalid email or password' : 'Invalid username or password');
    }

    // Generate tokens
    const tokenPayload: JWTPayload = { userId: user.id, email: user.email };
    const accessToken = generateAccessToken(tokenPayload);
    const refreshToken = generateRefreshToken(tokenPayload);

    return {
      user: {
        id: user.id,
        name: user.name,
        username: user.username,
        email: user.email,
      },
      accessToken,
      refreshToken,
    };
  }

  async refreshToken(refreshToken: string) {
    const decoded = verifyRefreshToken(refreshToken);
    if (!decoded) {
      throw new Error('Invalid refresh token');
    }

    const user = await prisma.user.findUnique({
      where: { id: decoded.userId },
      select: {
        id: true,
        email: true,
      }
    });

    if (!user) {
      throw new Error('User not found');
    }

    const tokenPayload: JWTPayload = { userId: user.id, email: user.email };
    const newAccessToken = generateAccessToken(tokenPayload);
    const newRefreshToken = generateRefreshToken(tokenPayload);

    return {
      accessToken: newAccessToken,
      refreshToken: newRefreshToken,
    };
  }

  async logout(refreshToken?: string) {
    if (refreshToken) {
      try {
        const decoded = verifyRefreshToken(refreshToken);
        if (decoded) {
          // In a real app, you might want to blacklist the token
          // For now, we'll just let it expire naturally
        }
      } catch (error) {
        // Token was invalid, but that's okay for logout
      }
    }
  }
}

File: backend/src/main/java/com/taskmanager/service/AuthService.java
package com.taskmanager.service;

import com.taskmanager.dto.AuthResponse;
import com.taskmanager.dto.RegisterRequest;
import com.taskmanager.model.User;
import com.taskmanager.repository.UserRepository;
import com.taskmanager.security.JwtTokenProvider;
import io.jsonwebtoken.Claims;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider tokenProvider;

    public AuthService(UserRepository userRepository, PasswordEncoder passwordEncoder, JwtTokenProvider tokenProvider) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.tokenProvider = tokenProvider;
    }

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        // Check email existence
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new IllegalArgumentException("Email already exists");
        }

        // Check username existence
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new IllegalArgumentException("Username already exists");
        }

        // Create new User
        User user = new User();
        user.setName(request.getName());
        user.setUsername(request.getUsername());
        user.setEmail(request.getEmail());
        user.setPassword(passwordEncoder.encode(request.getPassword()));

        User savedUser = userRepository.save(user);

        // Generate tokens
        String accessToken = tokenProvider.generateAccessToken(savedUser.getId(), savedUser.getEmail());
        String refreshToken = tokenProvider.generateRefreshToken(savedUser.getId(), savedUser.getEmail());

        AuthResponse.UserResponseUser userResponse = new AuthResponse.UserResponseUser(
                savedUser.getId(), savedUser.getName(), savedUser.getUsername(), savedUser.getEmail()
        );

        return new AuthResponse(userResponse, accessToken, refreshToken);
    }

    @Transactional(readOnly = true)
    public AuthResponse login(String loginIdentifier, String password) {
        boolean isEmail = loginIdentifier.contains("@");
        User user;

        if (isEmail) {
            user = userRepository.findByEmail(loginIdentifier)
                    .orElseThrow(() -> new IllegalArgumentException("Invalid email or password"));
        } else {
            user = userRepository.findByUsername(loginIdentifier)
                    .orElseThrow(() -> new IllegalArgumentException("Invalid username or password"));
        }

        if (!passwordEncoder.matches(password, user.getPassword())) {
            throw new IllegalArgumentException(isEmail ? "Invalid email or password" : "Invalid username or password");
        }

        // Generate tokens
        String accessToken = tokenProvider.generateAccessToken(user.getId(), user.getEmail());
        String refreshToken = tokenProvider.generateRefreshToken(user.getId(), user.getEmail());

        AuthResponse.UserResponseUser userResponse = new AuthResponse.UserResponseUser(
                user.getId(), user.getName(), user.getUsername(), user.getEmail()
        );

        return new AuthResponse(userResponse, accessToken, refreshToken);
    }

    @Transactional(readOnly = true)
    public AuthResponse refreshToken(String refreshToken) {
        Claims claims = tokenProvider.verifyRefreshToken(refreshToken);
        if (claims == null) {
            throw new IllegalArgumentException("Invalid refresh token");
        }

        Integer userId = claims.get("userId", Integer.class);
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));

        // Generate new tokens
        String newAccessToken = tokenProvider.generateAccessToken(user.getId(), user.getEmail());
        String newRefreshToken = tokenProvider.generateRefreshToken(user.getId(), user.getEmail());

        // We wrap it in AuthResponse but only accessToken/refreshToken are used
        return new AuthResponse(null, newAccessToken, newRefreshToken);
    }
}
